File: packages/ts/tracing/src/langchain.ts
import {
  Lemma,
  type LemmaClientOptions,
  type SpanHandle,
  type TraceHandle,
} from "./client";

type RunId = string;

type Serialized = {
  id?: string[];
  name?: string;
  kwargs?: Record<string, unknown>;
  [key: string]: unknown;
};

type LLMResult = {
  generations?: unknown[][];
  llmOutput?: Record<string, unknown>;
  [key: string]: unknown;
};

export type LangChainIntegrationOptions = {
  lemma?: Lemma;
  apiKey?: LemmaClientOptions["apiKey"];
  projectId?: LemmaClientOptions["projectId"];
  baseUrl?: LemmaClientOptions["baseUrl"];
  fetch?: LemmaClientOptions["fetch"];
  agentName?: string;
  metadata?: Record<string, unknown>;
  recordInputs?: boolean;
  recordOutputs?: boolean;
};

type StoredRun = {
  handle?: SpanHandle;
  trace?: TraceHandle;
  parentRunId?: string;
  type: "chain" | "llm" | "tool" | "retriever";
};

function serializedName(serialized: Serialized | undefined, fallback: string) {
  if (typeof serialized?.name === "string" && serialized.name) {
    return serialized.name;
  }
  const id = serialized?.id;
  if (Array.isArray(id) && id.length > 0) {
    return String(id[id.length - 1]);
  }
  return fallback;
}

function modelName(serialized: Serialized | undefined) {
  const kwargs = serialized?.kwargs;
  const value =
    kwargs?.model ??
    kwargs?.modelName ??
    kwargs?.model_name ??
    kwargs?.model_id ??
    serialized?.model ??
    serialized?.modelName ??
    serialized?.model_name ??
    serialized?.model_id;
  return typeof value === "string" ? value : undefined;
}

function firstText(value: unknown): string | undefined {
  if (typeof value === "string") return value;
  if (!value || typeof value !== "object") return undefined;
  const record = value as Record<string, unknown>;
  if (typeof record.text === "string") return record.text;
  if (typeof record.content === "string") return record.content;
  const message = record.message;
  if (message && typeof message === "object") {
    return firstText(message);
  }
  return undefined;
}

function llmOutput(result: LLMResult): unknown {
  const generations = result.generations;
  if (!Array.isArray(generations)) return result;
  const text = generations.flat().map(firstText).filter(Boolean).join("");
  return text || generations;
}

function errorMessage(error: unknown) {
  return error instanceof Error ? error.message : String(error);
}

export class LemmaLangChainCallbackHandler {
  name = "lemma";
  private lemma: Lemma | undefined;
  private readonly runs = new Map<RunId, StoredRun>();

  constructor(private readonly options: LangChainIntegrationOptions = {}) {
    this.lemma = options.lemma;
  }

  private getLemma() {
    this.lemma ??= new Lemma({
      apiKey: this.options.apiKey,
      projectId: this.options.projectId,
      baseUrl: this.options.baseUrl,
      fetch: this.options.fetch,
    });
    return this.lemma;
  }

  private traceName(serialized: Serialized | undefined, fallback: string) {
    return this.options.agentName ?? serializedName(serialized, fallback);
  }

  private startTrace(
    runId: RunId,
    serialized: Serialized | undefined,
    input: unknown,
    fallbackName: string,
    metadata?: Record<string, unknown>,
  ) {
    const trace = this.getLemma().trace({
      name: this.traceName(serialized, fallbackName),
      input: this.options.recordInputs === false ? undefined : input,
      metadata: {
        ...this.options.metadata,
        ...(metadata ?? {}),
        langchainRunId: runId,
      },
    });
    this.runs.set(runId, { trace, type: "chain" });
    return trace;
  }

  private parent(runId: RunId | undefined) {
    if (!runId) return undefined;
    return this.runs.get(runId);
  }

  handleChainStart(
    serialized: Serialized,
    inputs: unknown,
    runId: RunId,
    parentRunId?: RunId,
    _tags?: string[],
    metadata?: Record<string, unknown>,
    _runType?: string,
    name?: string,
  ) {
    const parent = this.parent(parentRunId);
    if (!parent) {
      this.startTrace(
        runId,
        { ...serialized, name: name ?? serialized?.name },
        inputs,
        "langchain-run",
        metadata,
      );
      return;
    }
    const handle = (parent.handle ?? parent.trace)?.startSpan({
      name: name ?? serializedName(serialized, "langchain-chain"),
      input: this.options.recordInputs === false ? undefined : inputs,
      metadata: this.options.metadata,
      attributes: {
        "langchain.run_id": runId,
        "langchain.parent_run_id": parentRunId,
        "langchain.run_type": "chain",
      },
    });
    this.runs.set(runId, { handle, parentRunId, type: "chain" });
  }

  async handleChainEnd(outputs: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    if (!run) return;
    if (run.trace) {
      await run.trace.end(
        this.options.recordOutputs === false ? undefined : { output: outputs },
      );
      this.runs.delete(runId);
      return;
    }
    run.handle?.end({
      output: this.options.recordOutputs === false ? undefined : outputs,
    });
    this.runs.delete(runId);
  }

  async handleChainError(error: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    if (!run) return;
    if (run.trace) {
      run.trace.fail(error);
      await run.trace.end();
      this.runs.delete(runId);
      return;
    }
    run.handle?.end({ status: "ERROR", error: errorMessage(error) });
    this.runs.delete(runId);
  }

  handleLLMStart(
    serialized: Serialized,
    prompts: string[],
    runId: RunId,
    parentRunId?: RunId,
    extraParams?: Record<string, unknown>,
  ) {
    const parent = this.parent(parentRunId);
    const trace =
      parent?.trace ??
      parent?.handle ??
      this.startTrace(runId, serialized, prompts, "langchain-llm");
    const handle = trace.startGeneration({
      name: serializedName(serialized, "langchain-llm"),
      input: this.options.recordInputs === false ? undefined : prompts,
      metadata: this.options.metadata,
      model: modelName(serialized),
      llmProvider: "langchain",
      llmInputMessages:
        this.options.recordInputs === false
          ? undefined
          : prompts.map((content) => ({ role: "user", content })),
      llmInvocationParameters: extraParams,
      attributes: {
        "langchain.run_id": runId,
        "langchain.parent_run_id": parentRunId,
        "langchain.run_type": "llm",
      },
    });
    this.runs.set(runId, { handle, parentRunId, type: "llm" });
  }

  handleChatModelStart(
    serialized: Serialized,
    messages: unknown[][],
    runId: RunId,
    parentRunId?: RunId,
    extraParams?: Record<string, unknown>,
  ) {
    const flatMessages = messages.flat();
    const parent = this.parent(parentRunId);
    const trace =
      parent?.trace ??
      parent?.handle ??
      this.startTrace(runId, serialized, flatMessages, "langchain-chat-model");
    const handle = trace.startGeneration({
      name: serializedName(serialized, "langchain-chat-model"),
      input: this.options.recordInputs === false ? undefined : flatMessages,
      metadata: this.options.metadata,
      model: modelName(serialized),
      llmProvider: "langchain",
      llmInputMessages:
        this.options.recordInputs === false ? undefined : flatMessages,
      llmInvocationParameters: extraParams,
      attributes: {
        "langchain.run_id": runId,
        "langchain.parent_run_id": parentRunId,
        "langchain.run_type": "llm",
      },
    });
    this.runs.set(runId, { handle, parentRunId, type: "llm" });
  }

  handleLLMEnd(output: LLMResult, runId: RunId) {
    const run = this.runs.get(runId);
    if (!run?.handle) return;
    const outputText = llmOutput(output);
    run.handle.end({
      output: this.options.recordOutputs === false ? undefined : outputText,
      llmOutputMessages:
        this.options.recordOutputs === false || outputText === undefined
          ? undefined
          : [{ role: "assistant", content: outputText }],
    });
    this.runs.delete(runId);
  }

  handleLLMError(error: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    run?.handle?.end({ status: "ERROR", error: errorMessage(error) });
    this.runs.delete(runId);
  }

  handleToolStart(
    serialized: Serialized,
    input: unknown,
    runId: RunId,
    parentRunId?: RunId,
  ) {
    const parent = this.parent(parentRunId);
    const trace =
      parent?.trace ??
      parent?.handle ??
      this.startTrace(runId, serialized, input, "langchain-tool");
    const name = serializedName(serialized, "langchain-tool");
    const handle = trace.startTool({
      name,
      toolName: name,
      input: this.options.recordInputs === false ? undefined : input,
      metadata: this.options.metadata,
      attributes: {
        "langchain.run_id": runId,
        "langchain.parent_run_id": parentRunId,
        "langchain.run_type": "tool",
      },
    });
    this.runs.set(runId, { handle, parentRunId, type: "tool" });
  }

  handleToolEnd(output: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    run?.handle?.end({
      output: this.options.recordOutputs === false ? undefined : output,
    });
    this.runs.delete(runId);
  }

  handleToolError(error: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    run?.handle?.end({ status: "ERROR", error: errorMessage(error) });
    this.runs.delete(runId);
  }

  handleRetrieverStart(
    serialized: Serialized,
    query: string,
    runId: RunId,
    parentRunId?: RunId,
  ) {
    const parent = this.parent(parentRunId);
    const trace =
      parent?.trace ??
      parent?.handle ??
      this.startTrace(runId, serialized, query, "langchain-retriever");
    const handle = trace.startSpan({
      name: serializedName(serialized, "langchain-retriever"),
      input: this.options.recordInputs === false ? undefined : query,
      metadata: this.options.metadata,
      attributes: {
        "langchain.run_id": runId,
        "langchain.parent_run_id": parentRunId,
        "langchain.run_type": "retriever",
      },
    });
    this.runs.set(runId, { handle, parentRunId, type: "retriever" });
  }

  handleRetrieverEnd(documents: unknown[], runId: RunId) {
    const run = this.runs.get(runId);
    run?.handle?.end({
      output: this.options.recordOutputs === false ? undefined : documents,
    });
    this.runs.delete(runId);
  }

  handleRetrieverError(error: unknown, runId: RunId) {
    const run = this.runs.get(runId);
    run?.handle?.end({ status: "ERROR", error: errorMessage(error) });
    this.runs.delete(runId);
  }
}

export const LangChainCallbackHandler = LemmaLangChainCallbackHandler;

export function langChain(options: LangChainIntegrationOptions = {}) {
  return new LemmaLangChainCallbackHandler(options);
}

export function langGraph(options: LangChainIntegrationOptions = {}) {
  return langChain({ agentName: "langgraph-agent", ...options });
}

export type LemmaLangChainIntegrationOptions = LangChainIntegrationOptions;
export type LemmaLangGraphIntegrationOptions = LangChainIntegrationOptions;

File: packages/py/tracing/src/uselemma_tracing/langchain.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

from .client import Lemma, SpanHandle, TraceContext, _now


@dataclass
class _Run:
    kind: str
    parent_run_id: str | None = None
    handle: SpanHandle | None = None
    trace: TraceContext | None = None
    started_at: datetime | None = None


def _get(obj: Any, key: str, default: Any = None) -> Any:
    if isinstance(obj, dict):
        return obj.get(key, default)
    return getattr(obj, key, default)


def _serialized_name(serialized: Any, fallback: str) -> str:
    name = _get(serialized, "name")
    if isinstance(name, str) and name:
        return name
    ids = _get(serialized, "id")
    if isinstance(ids, list) and ids:
        return str(ids[-1])
    return fallback


def _model_name(serialized: Any) -> str | None:
    kwargs = _get(serialized, "kwargs", {}) or {}
    for source in (kwargs, serialized):
        for key in ("model", "model_name", "modelName", "model_id", "modelId"):
            value = _get(source, key)
            if isinstance(value, str):
                return value
    return None


def _first_text(value: Any) -> str | None:
    if isinstance(value, str):
        return value
    if isinstance(value, dict):
        if isinstance(value.get("text"), str):
            return value["text"]
        if isinstance(value.get("content"), str):
            return value["content"]
        return _first_text(value.get("message"))
    text = getattr(value, "text", None)
    if isinstance(text, str):
        return text
    content = getattr(value, "content", None)
    if isinstance(content, str):
        return content
    message = getattr(value, "message", None)
    if message is not None:
        return _first_text(message)
    return None


def _llm_output(response: Any) -> Any:
    generations = _get(response, "generations")
    if not isinstance(generations, list):
        return response
    text = "".join(
        text
        for group in generations
        if isinstance(group, list)
        for item in group
        if (text := _first_text(item))
    )
    return text or generations



def _error_message(error: Any) -> str:
    return str(error)


class LemmaLangChainCallbackHandler:
    name = "lemma"

    def __init__(
        self,
        lemma: Lemma | None = None,
        *,
        api_key: str | None = None,
        project_id: str | None = None,
        base_url: str = "https://api.uselemma.ai",
        transport: Any = None,
        agent_name: str | None = None,
        metadata: dict[str, Any] | None = None,
        record_inputs: bool = True,
        record_outputs: bool = True,
    ) -> None:
        self.lemma = lemma or Lemma(
            api_key=api_key,
            project_id=project_id,
            base_url=base_url,
            transport=transport,
        )
        self.agent_name = agent_name
        self.metadata = metadata or {}
        self.record_inputs = record_inputs
        self.record_outputs = record_outputs
        self._runs: dict[str, _Run] = {}

    def _trace_name(self, serialized: Any, fallback: str) -> str:
        return self.agent_name or _serialized_name(serialized, fallback)

    def _start_trace(
        self,
        run_id: str,
        serialized: Any,
        input: Any,
        fallback_name: str,
        metadata: dict[str, Any] | None = None,
    ) -> TraceContext:
        ctx = TraceContext(
            name=self._trace_name(serialized, fallback_name),
            input=input if self.record_inputs else None,
            metadata={
                **self.metadata,
                **(metadata or {}),
                "langchain_run_id": str(run_id),
            },
        )
        self._runs[str(run_id)] = _Run(kind="chain", trace=ctx, started_at=_now())
        return ctx

    def _parent(self, parent_run_id: str | None) -> _Run | None:
        if parent_run_id is None:
            return None
        return self._runs.get(str(parent_run_id))

    def _parent_target(self, parent_run_id: str | None) -> TraceContext | SpanHandle | None:
        parent = self._parent(parent_run_id)
        if parent is None:
            return None
        return parent.handle or parent.trace

    def _start_span(
        self, parent: TraceContext | SpanHandle, **kwargs: Any
    ) -> SpanHandle:
        if isinstance(parent, SpanHandle):
            return parent.trace.start_span(parent_id=parent.id, **kwargs)
        return parent.start_span(**kwargs)

    def _start_generation(
        self, parent: TraceContext | SpanHandle, **kwargs: Any
    ) -> SpanHandle:
        if isinstance(parent, SpanHandle):
            return parent.trace.start_generation(parent_id=parent.id, **kwargs)
        return parent.start_generation(**kwargs)

    def _start_tool(
        self, parent: TraceContext | SpanHandle, **kwargs: Any
    ) -> SpanHandle:
        if isinstance(parent, SpanHandle):
            return parent.trace.start_tool(parent_id=parent.id, **kwargs)
        return parent.start_tool(**kwargs)

    def on_chain_start(
        self,
        serialized: Any,
        inputs: Any,
        *,
        run_id: str,
        parent_run_id: str | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        run_type: str | None = None,
        name: str | None = None,
        **_: Any,
    ) -> None:
        parent = self._parent_target(parent_run_id)
        if parent is None:
            trace_serialized = dict(serialized or {})
            if name:
                trace_serialized["name"] = name
            self._start_trace(run_id, trace_serialized, inputs, "langchain-run", metadata)
            return
        handle = self._start_span(
            parent,
            name=name or _serialized_name(serialized, "langchain-chain"),
            input=inputs if self.record_inputs else None,
            metadata=self.metadata,
            attributes={
                "langchain.run_id": str(run_id),
                "langchain.parent_run_id": str(parent_run_id),
                "langchain.run_type": run_type or "chain",
            },
        )
        self._runs[str(run_id)] = _Run(
            kind="chain", parent_run_id=str(parent_run_id), handle=handle
        )

    def on_chain_end(self, outputs: Any, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        if run is None:
            return
        if run.trace is not None:
            if self.record_outputs:
                run.trace.output(outputs)
            self.lemma._send(run.trace, run.started_at or _now(), _now())
            return
        run.handle and run.handle.end(output=outputs if self.record_outputs else None)

    def on_chain_error(self, error: BaseException, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        if run is None:
            return
        if run.trace is not None:
            run.trace.fail(error)
            self.lemma._send(run.trace, run.started_at or _now(), _now())
            return
        run.handle and run.handle.end(status="ERROR", error=_error_message(error))

    def on_llm_start(
        self,
        serialized: Any,
        prompts: list[str],
        *,
        run_id: str,
        parent_run_id: str | None = None,
        invocation_params: dict[str, Any] | None = None,
        **_: Any,
    ) -> None:
        parent = self._parent_target(parent_run_id) or self._start_trace(
            run_id, serialized, prompts, "langchain-llm"
        )
        handle = self._start_generation(
            parent,
            name=_serialized_name(serialized, "langchain-llm"),
            input=prompts if self.record_inputs else None,
            metadata=self.metadata,
            model=_model_name(serialized),
            llm_provider="langchain",
            llm_input_messages=(
                [{"role": "user", "content": prompt} for prompt in prompts]
                if self.record_inputs
                else None
            ),
            llm_invocation_parameters=invocation_params,
            attributes={
                "langchain.run_id": str(run_id),
                "langchain.parent_run_id": str(parent_run_id)
                if parent_run_id is not None
                else None,
                "langchain.run_type": "llm",
            },
        )
        self._runs[str(run_id)] = _Run(
            kind="llm",
            parent_run_id=str(parent_run_id) if parent_run_id is not None else None,
            handle=handle,
        )

    def on_chat_model_start(
        self,
        serialized: Any,
        messages: list[list[Any]],
        *,
        run_id: str,
        parent_run_id: str | None = None,
        invocation_params: dict[str, Any] | None = None,
        **_: Any,
    ) -> None:
        flat_messages = [message for group in messages for message in group]
        parent = self._parent_target(parent_run_id) or self._start_trace(
            run_id, serialized, flat_messages, "langchain-chat-model"
        )
        handle = self._start_generation(
            parent,
            name=_serialized_name(serialized, "langchain-chat-model"),
            input=flat_messages if self.record_inputs else None,
            metadata=self.metadata,
            model=_model_name(serialized),
            llm_provider="langchain",
            llm_input_messages=flat_messages if self.record_inputs else None,
            llm_invocation_parameters=invocation_params,
            attributes={
                "langchain.run_id": str(run_id),
                "langchain.parent_run_id": str(parent_run_id)
                if parent_run_id is not None
                else None,
                "langchain.run_type": "llm",
            },
        )
        self._runs[str(run_id)] = _Run(
            kind="llm",
            parent_run_id=str(parent_run_id) if parent_run_id is not None else None,
            handle=handle,
        )

    def on_llm_end(self, response: Any, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        if run is None or run.handle is None:
            return
        output = _llm_output(response)
        run.handle.end(
            output=output if self.record_outputs else None,
            llm_output_messages=(
                [{"role": "assistant", "content": output}]
                if self.record_outputs and output is not None
                else None
            ),
        )

    def on_llm_error(self, error: BaseException, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        run and run.handle and run.handle.end(status="ERROR", error=_error_message(error))

    def on_tool_start(
        self,
        serialized: Any,
        input_str: Any,
        *,
        run_id: str,
        parent_run_id: str | None = None,
        **_: Any,
    ) -> None:
        parent = self._parent_target(parent_run_id) or self._start_trace(
            run_id, serialized, input_str, "langchain-tool"
        )
        name = _serialized_name(serialized, "langchain-tool")
        handle = self._start_tool(
            parent,
            name=name,
            tool_name=name,
            input=input_str if self.record_inputs else None,
            metadata=self.metadata,
            attributes={
                "langchain.run_id": str(run_id),
                "langchain.parent_run_id": str(parent_run_id)
                if parent_run_id is not None
                else None,
                "langchain.run_type": "tool",
            },
        )
        self._runs[str(run_id)] = _Run(
            kind="tool",
            parent_run_id=str(parent_run_id) if parent_run_id is not None else None,
            handle=handle,
        )

    def on_tool_end(self, output: Any, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        run and run.handle and run.handle.end(
            output=output if self.record_outputs else None
        )

    def on_tool_error(self, error: BaseException, *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        run and run.handle and run.handle.end(
            status="ERROR", error=_error_message(error)
        )

    def on_retriever_start(
        self,
        serialized: Any,
        query: str,
        *,
        run_id: str,
        parent_run_id: str | None = None,
        **_: Any,
    ) -> None:
        parent = self._parent_target(parent_run_id) or self._start_trace(
            run_id, serialized, query, "langchain-retriever"
        )
        handle = self._start_span(
            parent,
            name=_serialized_name(serialized, "langchain-retriever"),
            input=query if self.record_inputs else None,
            metadata=self.metadata,
            attributes={
                "langchain.run_id": str(run_id),
                "langchain.parent_run_id": str(parent_run_id)
                if parent_run_id is not None
                else None,
                "langchain.run_type": "retriever",
            },
        )
        self._runs[str(run_id)] = _Run(
            kind="retriever",
            parent_run_id=str(parent_run_id) if parent_run_id is not None else None,
            handle=handle,
        )

    def on_retriever_end(self, documents: list[Any], *, run_id: str, **_: Any) -> None:
        run = self._runs.pop(str(run_id), None)
        run and run.handle and run.handle.end(
            output=documents if self.record_outputs else None,
        )

    def on_retriever_error(
        self, error: BaseException, *, run_id: str, **_: Any
    ) -> None:
        run = self._runs.pop(str(run_id), None)
        run and run.handle and run.handle.end(
            status="ERROR", error=_error_message(error)
        )


def langchain(**options: Any) -> LemmaLangChainCallbackHandler:
    return LemmaLangChainCallbackHandler(**options)


def langgraph(**options: Any) -> LemmaLangChainCallbackHandler:
    return LemmaLangChainCallbackHandler(agent_name="langgraph-agent", **options)
